File: 김태연/백준/Gold/20057. 마법사 상어와 토네이도/마법사 상어와 토네이도.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    static int N;
    static int[][] map;
    // 좌,하,우,상
    static int[] dr = { 0, 1, 0, -1 };
    static int[] dc = { -1, 0, 1, 0 };

    // tdr 다시 확인하기
    static int[][] tdr = { { 0, 1, -1, 2, 1, -1, -2, 1, -1, 0},
            { 2, 1, 1, 0, 0, 0, 0, -1, -1, 1 },
            { 0, -1, 1, -2, -1, 1, 2, -1, 1, 0 },
            { -2, -1, -1, 0, 0, 0, 0, 1, 1, -1 } }; // 여기까지 토네이도 위치

    static int[][] tdc = { { -2, -1, -1, 0, 0, 0, 0, 1, 1, -1 }, { 0, -1, 1, -2, -1, 1, 2, -1, 1, 0 },
            { 2, 1, 1, 0, 0, 0, 0, -1, -1, 1 }, { 0, -1, 1, -2, -1, 1, 2, -1, 1, 0 } };

    static int[] tPercent = { 5, 10, 10, 2, 7, 7, 2, 1, 1, 0 };

    static int dir = 0;
    static int sum = 0;
    static int binaryCount = 0;
    static int L;
    static int moveCount;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        map = new int[N][N];
        sum = 0;

        // 토네이도의 시작 지점 (r,c)
        int r = N / 2;
        int c = N / 2;
        L = 1;
        moveCount = 1;
        
        for (int i = 0; i < N; i++) {
            String[] input = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                map[i][j] = Integer.parseInt(input[j]);
            }
        }
        
        while (!(r == 0 && c == 0)) {

            // 1. 주어진 좌표와 방향에서 모래를 흩뿌린다
            spread(r, c, dir);
            moveCount--;

            // 2. 이동 방향으로 토네이도 옮겨놓고
            r = r + dr[dir];
            c = c + dc[dir];
            
            // 3. 다음 방향 계산
            if (moveCount == 0) {
                changeDirection();
            }
        }

        System.out.println(sum);
    }

    // 토네이도의 위치를 x
    static void spread(int r, int c, int dir) {

        int nextR = r + dr[dir];
        int nextC = c + dc[dir];
        
        if (!isBoundary(nextR,nextC)) return;
        
        int sand = map[nextR][nextC];

        for (int i = 0; i < 10; i++) {
            // 이번에 확인할 row 위치
            int row = nextR + tdr[dir][i];

            // 이번에 확인할 column 위치
            int column = nextC + tdc[dir][i];

            // 이번에 움직일 모래의 양
            int sandAmount = (int) (sand * 0.01 * tPercent[i]);

            // 마지막 회차때는
            if (i == 9)
                sandAmount = map[nextR][nextC];

            if (isBoundary(row, column)) {
                map[row][column] += sandAmount;
                map[nextR][nextC] -= sandAmount;
                // 모래 처리 로직
                // 알파 로직 처리
            } else {
                sum += sandAmount;
                map[nextR][nextC] -= sandAmount;
            }

        }

    }

    static boolean isBoundary(int row, int column) {
    	 return row >= 0 && row < N && column >= 0 && column < N;
    }

    static void changeDirection() {
        binaryCount++;
        dir = (dir + 1) % 4;
        if (binaryCount % 2 == 0) {
            L++;
        }
        moveCount = L;
    }

}

File: 이준영/백준/Gold/20057. 마법사 상어와 토네이도/마법사 상어와 토네이도.py
# 처음엔 딕셔너리로 관리하려고 하였으나 실수를 하기 쉽다고 판단하여 배열로 관리하였습니다. (토네이도로 인한 변화를)
# 실수계산이 느리니 자연수로 연산을 하도록 하였습니다.

N = int(input())
arr = [list(map(int, input().split())) for _ in range(N)]

dy = [0, 1, 0, -1]                                                  # 좌, 하, 우, 상
dx = [-1, 0, 1, 0]
tmp = [[0, 0, 2, 0, 0],
       [0, 10, 7, 1, 0],
       [5, 0, 0, 0, 0],
       [0, 10, 7, 1, 0],
       [0, 0, 2, 0, 0]]

sand = [0] * 4
sand[0] = [row[:] for row in tmp]                                   # sand[d]는 d방향으로 바람이 불 때 변화량입니다.
                                                                    # tmp배열을 반시계 방향으로 회전하며 sand배열안에 넣었습니다.
for l in range(1, 4):
    b = [[0] * 5 for _ in range(5)]
    for i in range(5):
        for j in range(5):
            b[i][j] = tmp[j][4 - i]
    sand[l] = b
    tmp = b

d = 0                                                               # 현재 방향
out = 0                                                             # 밖으로 나간 모래 양
y = x = N // 2                                                      # 토네이도의 위치

for l in range(1, N):                                               # 달팽이를 생각해보자 앞으로 전진하는 거리는
    for _ in range(2):                                              # 1,1,2,2,3,3 .... N-1,N-1 에 N-1한번 더 이다.
        for t in range(l):                                          # 각 직진이 끝날 때마다 반시계방향으로 회전한다.
            y = y + dy[d]                                            
            x = x + dx[d]
            tot = arr[y][x]                                         # 현재 칸의 모래 양
            for i in range(5):
                for j in range(5):
                    if sand[d][i][j] == 0:                          # 0인 곳은 모래가 퍼지지 않는다.
                        continue
                    else:
                        tmp = arr[y][x] * sand[d][i][j] // 100      # 모래가 퍼져나가는 양
                        ny = y + i - 2                              # (2,2)를 기준으로 y,x 좌표 변화량을 세야 한다.
                        nx = x + j - 2
                        if 0 <= ny < N and 0 <= nx < N:             # 범위 안에 있으면 더하고
                            arr[ny][nx] += tmp
                        else:
                            out += tmp                              # 아니면 밖으로 나간다.
                        tot -= tmp                                  # 남은 모래 양을 줄인다.

            arr[y][x] = 0                                           # 현재 칸 비워주고
            ny = y + dy[d]                                          # 알파 칸을 보자
            nx = x + dx[d]
            if 0 <= ny < N and 0 <= nx < N:                         # 만약 범위내에 있으면
                arr[ny][nx] += tot                                  # 남은 모래를 다 준다.
            else:
                out += tot                                          # 아니면 바깥에 기부

        d = (d + 1) % 4                                             # 방향 전환

for t in range(N - 1):                                              # 마지막 왼쪽으로 N-1번 전진
    y = y + dy[d]
    x = x + dx[d]
    tot = arr[y][x]
    for i in range(5):
        for j in range(5):
            if sand[d][i][j] == 0:
                continue
            else:
                tmp = arr[y][x] * sand[d][i][j] // 100
                ny = y + i - 2
                nx = x + j - 2
                if 0 <= ny < N and 0 <= nx < N:
                    arr[ny][nx] += tmp
                else:
                    out += tmp
                tot -= tmp

    arr[y][x] = 0
    ny = y + dy[d]
    nx = x + dx[d]
    if 0 <= ny < N and 0 <= nx < N:
        arr[ny][nx] += tot
    else:
        out += tot

print(out)
